File: Best_Research/#Done/Rust/LearningRust/Units/Physics/src/units/time/display.rs
use std::fmt;

use crate::internal::format_unit_value;

use super::quantity::{Time, TimeUnit};

impl Time
{
    #[must_use]
    pub(crate) const fn display_as(self, unit: TimeUnit) -> TimeDisplay
    {
        TimeDisplay { value: self, unit, precision: None }
    }

    #[must_use]
    pub(crate) const fn display_as_precision(self, unit: TimeUnit, precision: usize) -> TimeDisplay
    {
        TimeDisplay {
            value: self,
            unit,
            precision: Some(precision),
        }
    }

    #[must_use]
    pub const fn display_seconds(self) -> TimeDisplay
    {
        self.display_as(TimeUnit::Seconds)
    }

    #[must_use]
    pub const fn display_milliseconds(self) -> TimeDisplay
    {
        self.display_as(TimeUnit::Milliseconds)
    }

    #[must_use]
    pub const fn display_microseconds(self) -> TimeDisplay
    {
        self.display_as(TimeUnit::Microseconds)
    }

    #[must_use]
    pub const fn display_nanoseconds(self) -> TimeDisplay
    {
        self.display_as(TimeUnit::Nanoseconds)
    }

    #[must_use]
    pub const fn display_minutes(self) -> TimeDisplay
    {
        self.display_as(TimeUnit::Minutes)
    }

    #[must_use]
    pub const fn display_hours(self) -> TimeDisplay
    {
        self.display_as(TimeUnit::Hours)
    }

    #[must_use]
    pub const fn display_seconds_precision(self, precision: usize) -> TimeDisplay
    {
        self.display_as_precision(TimeUnit::Seconds, precision)
    }

    #[must_use]
    pub const fn display_milliseconds_precision(self, precision: usize) -> TimeDisplay
    {
        self.display_as_precision(TimeUnit::Milliseconds, precision)
    }

    #[must_use]
    pub const fn display_microseconds_precision(self, precision: usize) -> TimeDisplay
    {
        self.display_as_precision(TimeUnit::Microseconds, precision)
    }

    #[must_use]
    pub const fn display_nanoseconds_precision(self, precision: usize) -> TimeDisplay
    {
        self.display_as_precision(TimeUnit::Nanoseconds, precision)
    }

    #[must_use]
    pub const fn display_minutes_precision(self, precision: usize) -> TimeDisplay
    {
        self.display_as_precision(TimeUnit::Minutes, precision)
    }

    #[must_use]
    pub const fn display_hours_precision(self, precision: usize) -> TimeDisplay
    {
        self.display_as_precision(TimeUnit::Hours, precision)
    }
}

#[derive(Copy, Clone, Debug)]
pub struct TimeDisplay
{
    value: Time,
    unit: TimeUnit,
    precision: Option<usize>,
}

impl fmt::Display for TimeDisplay
{
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result
    {
        let value = match self.unit
        {
            TimeUnit::Seconds => self.value.to_seconds(),
            TimeUnit::Milliseconds => self.value.to_milliseconds(),
            TimeUnit::Microseconds => self.value.to_microseconds(),
            TimeUnit::Nanoseconds => self.value.to_nanoseconds(),
            TimeUnit::Minutes => self.value.to_minutes(),
            TimeUnit::Hours => self.value.to_hours(),
        };

        format_unit_value(formatter, value, self.unit.symbol(), self.precision)
    }
}

impl fmt::Display for Time
{
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result
    {
        self.display_seconds().fmt(formatter)
    }
}

File: Best_Research/#Done/C++/Units-CppXmake/Physics/time/display.cpp
#include "physics/display/display.hpp"

#include <ostream>

namespace physics::time
{

    const char* symbol(TimeUnit unit)
    {
        switch (unit)
        {
        case TimeUnit::Seconds:
            return "s";
        case TimeUnit::Milliseconds:
            return "ms";
        case TimeUnit::Microseconds:
            return "us";
        case TimeUnit::Nanoseconds:
            return "ns";
        case TimeUnit::Minutes:
            return "min";
        case TimeUnit::Hours:
            return "h";
        }

        return "";
    }

    double value_in_unit(Time value, TimeUnit unit)
    {
        switch (unit)
        {
        case TimeUnit::Seconds:
            return value.to_seconds();
        case TimeUnit::Milliseconds:
            return value.to_milliseconds();
        case TimeUnit::Microseconds:
            return value.to_microseconds();
        case TimeUnit::Nanoseconds:
            return value.to_nanoseconds();
        case TimeUnit::Minutes:
            return value.to_minutes();
        case TimeUnit::Hours:
            return value.to_hours();
        }

        return value.to_seconds();
    }

    QuantityDisplay<Time, TimeUnit> Time::display_seconds()
    {
        return display_as(TimeUnit::Seconds);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_milliseconds()
    {
        return display_as(TimeUnit::Milliseconds);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_microseconds()
    {
        return display_as(TimeUnit::Microseconds);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_nanoseconds()
    {
        return display_as(TimeUnit::Nanoseconds);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_minutes()
    {
        return display_as(TimeUnit::Minutes);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_hours()
    {
        return display_as(TimeUnit::Hours);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_seconds_precision(int precision)
    {
        return display_as_precision(TimeUnit::Seconds, precision);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_milliseconds_precision(int precision)
    {
        return display_as_precision(TimeUnit::Milliseconds, precision);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_microseconds_precision(int precision)
    {
        return display_as_precision(TimeUnit::Microseconds, precision);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_nanoseconds_precision(int precision)
    {
        return display_as_precision(TimeUnit::Nanoseconds, precision);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_minutes_precision(int precision)
    {
        return display_as_precision(TimeUnit::Minutes, precision);
    }

    QuantityDisplay<Time, TimeUnit> Time::display_hours_precision(int precision)
    {
        return display_as_precision(TimeUnit::Hours, precision);
    }

    std::ostream& operator<<(std::ostream& stream, Time value)
    {
        return stream << value.display_as(TimeUnit::Seconds);
    }

} // namespace physics::time
